000X-B-C-I-R-S-error-handling-codes.md

```diff
@@ -0,0 +1,337 @@
+  "details": [
+    {
+      "@type": "type.googleapis.com/google.rpc.ErrorInfo",
+      "reason": "DPR_COMP_STS_1",
+      "domain": "redis.state.contrib.dapr.io",
+      "metadata": {
+        "key": "myapp||name"
+      }
+    },
+    {
+      "@type": "type.googleapis.com/google.rpc.ResourceInfo",
+      "resource_type": "StateStore",
+      "resource_name": "REDIS",
+      "owner": "",
+      "description": "unable to perform Redis.HGETALL call: dial tcp [::1]:6379: connect: connection refused"
+    }
+  ]
+}
+```
+
+
+### Sample Code Snippet (Go)
+```go
+import (
+   ...
+   "google.golang.org/genproto/googleapis/rpc/errdetails"
+   "google.golang.org/grpc/codes"
+   "google.golang.org/grpc/status"
+   ...
+)
+...
+if err != nil {
+  ...
+  ste := status.Newf(codes.Internal, messages.ErrStateGet, in.Key, in.StoreName, err.Error())
+  ei := errdetails.ErrorInfo{
+	 Domain: "grpc.runtime.dapr.io",
+         Reason: "DPR_RT_GRPC_1",
+	 Metadata: map[string]string{
+	       "key":       in.Key,
+	       "storeName": in.StoreName,
+	 },
+  }
+  ste, err2 := ste.WithDetails(&ei)
+  ...
+}
+```
+
+### Dapr Custom Error Info
+```go
+message DaprKitErrorInfo {
+  string reason = 1;
+  string domain = 2;
+  map<string, string> metadata = 3;
+}
+```
+
+### Example Dapr Custom Error Info
+```json
+{
+  "code": 8089,
+  "message": "unable to perform Redis.HGETALL call: dial tcp [::1]:6379: connect: connection refused",
+  "details": [
+    {
+      "@type": "type.googleapis.com/google.rpc.ErrorInfo",
+      "reason": "DPR_COMP_STS_1",
+      "domain": "redis.state.contrib.dapr.io",
+      "metadata": {
+        "key": "myapp||name"
+      }
+    },
+    {
+      "@type": "type.googleapis.com/google.rpc.ResourceInfo",
+      "resource_type": "StateStore",
+      "resource_name": "REDIS",
+      "owner": "",
+      "description": "unable to perform Redis.HGETALL call: dial tcp [::1]:6379: connect: connection refused"
+    },
+    {
+      "@type": "type.googleapis.com/kit.proto.customerrors.v1.DaprKitErrorInfo",
+      "reason": "DPR_COMP_STS_1",
+      "domain": "redis.state.contrib.dapr.io",
+      "metadata": {
+        "key": "myapp||name"
+      }
+    }
+  ]
+}
+```
+### Pros
+- Since the Dapr Runtime is using protocol buffers as the data format, support for the richer error model is already included in most of the gRPC implementations.
+- This would help minimize the changes with the Dapr ecosystem.
+- This solution could be used to programmatically react to errors as it provides a standard structure for the errors with details.
+
+### Cons
+- Dependencies on gPRC richer error model.
+- Not all gRPC implementations support this out of the box, [C#/dotnet](https://learn.microsoft.com/en-us/dotnet/architecture/grpc-for-wcf-developers/error-handling#grpc-richer-error-model) being one of them, but there is a [possible solution (Google.Api.CommonProtos)](https://www.nuget.org/packages/Google.Api.CommonProtos/) to this.
```

Comment:
This is a big deal, but we need to confirm with @halspang if the extra dependency is OK.

Reply:
`dotnet-sdk` works well. I need to test `js-sdk`, `java-sdk`, and `python-sdk`